File: examples/cosmos-evm/shared_secret.py
#!/usr/bin/env python3

import hashlib
import secrets
import binascii

def generate_secret_and_hashlock():
    """
    Generate a random secret and its corresponding hashlock
    Returns: Object containing secret and hashlock
    """
    # Generate a random 32-byte secret
    secret_bytes = secrets.token_bytes(32)
    secret_hex = '0x' + secret_bytes.hex()
    
    # Create hashlock (sha256 hash of the secret)
    hashlock = hashlib.sha256(secret_bytes).hexdigest()
    hashlock_hex = '0x' + hashlock
    
    return {
        'secret': secret_hex,
        'secret_bytes': secret_bytes,
        'hashlock': hashlock_hex,
        'hashlock_string': hashlock  # For Cosmos (string format)
    }

def validate_secret(secret, hashlock):
    """
    Validate if a secret matches a given hashlock
    Args:
        secret: The secret to validate
        hashlock: The hashlock to check against
    Returns: True if secret matches hashlock
    """
    # Remove 0x prefix if present
    clean_secret = secret[2:] if secret.startswith('0x') else secret
    secret_bytes = binascii.unhexlify(clean_secret)
    
    # Create hashlock from secret
    calculated_hashlock = hashlib.sha256(secret_bytes).hexdigest()
    expected_hashlock = hashlock[2:] if hashlock.startswith('0x') else hashlock
    
    return calculated_hashlock == expected_hashlock

def cosmos_to_evm_hashlock(cosmos_hashlock):
    """
    Convert Cosmos hashlock format to EVM format
    Args:
        cosmos_hashlock: Hashlock in Cosmos format (hex string)
    Returns: Hashlock in EVM format (0x...)
    """
    return cosmos_hashlock if cosmos_hashlock.startswith('0x') else '0x' + cosmos_hashlock

def evm_to_cosmos_hashlock(evm_hashlock):
    """
    Convert EVM hashlock format to Cosmos format
    Args:
        evm_hashlock: Hashlock in EVM format (0x...)
    Returns: Hashlock in Cosmos format (hex string without 0x)
    """
    return evm_hashlock[2:] if evm_hashlock.startswith('0x') else evm_hashlock

if __name__ == "__main__":
    print("Generating secret and hashlock for cross-chain HTCL...")
    result = generate_secret_and_hashlock()
    
    print("Generated values:")
    print("Secret (EVM):", result['secret'])
    print("Hashlock (EVM):", result['hashlock'])
    print("Hashlock (Cosmos):", result['hashlock_string'])
    
    print("\nValidation test:")
    is_valid = validate_secret(result['secret'], result['hashlock'])
    print("Secret validation:", is_valid)
    
    print("\nFormat conversion:")
    print("EVM to Cosmos:", evm_to_cosmos_hashlock(result['hashlock']))
    print("Cosmos to EVM:", cosmos_to_evm_hashlock(result['hashlock_string'])) 

File: examples/evm-cosmos/shared_secret.js
const crypto = require('crypto');

/**
 * Generate a random secret and its corresponding hashlock
 * @returns {Object} Object containing secret and hashlock
 */
function generateSecretAndHashlock() {
    // Generate a random 32-byte secret
    const secret = crypto.randomBytes(32);
    const secretHex = '0x' + secret.toString('hex');

    // Create hashlock (keccak256 hash of the secret)
    const hashlock = crypto.createHash('sha256').update(secret).digest('hex');
    const hashlockHex = '0x' + hashlock;

    return {
        secret: secretHex,
        secretBytes: secret,
        hashlock: hashlockHex,
        hashlockString: hashlock // For Cosmos (string format)
    };
}

/**
 * Validate if a secret matches a given hashlock
 * @param {string} secret - The secret to validate
 * @param {string} hashlock - The hashlock to check against
 * @returns {boolean} True if secret matches hashlock
 */
function validateSecret(secret, hashlock) {
    // Remove 0x prefix if present
    const cleanSecret = secret.startsWith('0x') ? secret.slice(2) : secret;
    const secretBytes = Buffer.from(cleanSecret, 'hex');

    // Create hashlock from secret
    const calculatedHashlock = crypto.createHash('sha256').update(secretBytes).digest('hex');
    const expectedHashlock = hashlock.startsWith('0x') ? hashlock.slice(2) : hashlock;

    return calculatedHashlock === expectedHashlock;
}

/**
 * Convert EVM hashlock format to Cosmos format
 * @param {string} evmHashlock - Hashlock in EVM format (0x...)
 * @returns {string} Hashlock in Cosmos format (hex string without 0x)
 */
function evmToCosmosHashlock(evmHashlock) {
    return evmHashlock.startsWith('0x') ? evmHashlock.slice(2) : evmHashlock;
}

/**
 * Convert Cosmos hashlock format to EVM format
 * @param {string} cosmosHashlock - Hashlock in Cosmos format (hex string)
 * @returns {string} Hashlock in EVM format (0x...)
 */
function cosmosToEvmHashlock(cosmosHashlock) {
    return cosmosHashlock.startsWith('0x') ? cosmosHashlock : '0x' + cosmosHashlock;
}

module.exports = {
    generateSecretAndHashlock,
    validateSecret,
    evmToCosmosHashlock,
    cosmosToEvmHashlock
};

// Example usage
if (require.main === module) {
    console.log('Generating secret and hashlock for cross-chain HTCL...');
    const { secret, hashlock, hashlockString } = generateSecretAndHashlock();

    console.log('Generated values:');
    console.log('Secret (EVM):', secret);
    console.log('Hashlock (EVM):', hashlock);
    console.log('Hashlock (Cosmos):', hashlockString);

    console.log('\nValidation test:');
    const isValid = validateSecret(secret, hashlock);
    console.log('Secret validation:', isValid);

    console.log('\nFormat conversion:');
    console.log('EVM to Cosmos:', evmToCosmosHashlock(hashlock));
    console.log('Cosmos to EVM:', cosmosToEvmHashlock(hashlockString));
} 